src/compiler/scala/tools/reflect/FormatInterpolator.scala

```diff
@@ -155,16 +162,24 @@ abstract class FormatInterpolator {
             val cv = Conversion(matches.next(), part0.pos, argc)
             if (n == 0 && cv.hasFlag('<')) cv.badFlag('<', "No last arg")
             else if (!cv.isLiteral && !cv.isIndexed) errorLeading(cv)
+            formatting = true
           }
           loop(more, n = n + 1)
         case Nil =>
       }
     }
     loop(parts, n = 0)
 
+    def constantly(s: String) = {
+      val k = Constant(s)
+      Literal(k).setType(ConstantType(k))
+    }
+
     //q"{..$evals; new StringOps(${fstring.toString}).format(..$ids)}"
     val format = amended.mkString
-    if (actuals.isEmpty && !format.contains("%")) Literal(Constant(format))
+    if (actuals.isEmpty && !formatting) constantly(format)
+    else if (!reported && actuals.forall(cond(_) { case Literal(c @ Constant(_)) => c.isSuitableLiteralType })) constantly(format.format(actuals.map(_.asInstanceOf[Literal].value.value).toIndexedSeq: _*))
```

Comment:
Similar issue here, but a bit worse: `Byte` and `Short` must be excluded as well, in addition to `Unit`, `Float` and `Double`. In theory `Byte` and `Short` should only be excluded for `%x`, `%X` and `%o`, while `Unit`, `Float` and `Double` should only be excluded for `%s`, but it's probably easier to excluded all of them all the time.